File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        long[] trees = new long[n];

        long max = 0;
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            trees[i] = Long.parseLong(st.nextToken());
            max = Math.max(trees[i], max);
        }

        long left = 0;
        long right = max;
        long result = 0;

        while (left <= right) {
            long mid = (left + right) / 2;
            long sum = 0;

            for (int i = 0; i < n; i++) {
                if (trees[i] > mid) {
                    sum += (trees[i] - mid);
                }
            }

            if (sum < m) {
                right = mid - 1;
            } else {
                result = mid;
                left = mid + 1;
            }
        }

        System.out.print(result);
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.kt
import java.util.*

fun main() {
    val br = System.`in`.bufferedReader()
    val (n, m) = br.readLine().split(' ').map { it.toLong() }
    val tree = StringTokenizer(br.readLine()).run { LongArray(n.toInt()) { nextToken().toLong() } }
    tree.sort()
    var en = tree[n.toInt() - 1]
    var st = 0L
    var mid = (st + en) / 2
    while (st <= en) {
        var sum = 0L
        tree.forEach {
            if (it >= mid) {
                sum += it - mid
            }
        }
        if (sum >= m) {
            st = mid + 1
        } else {
            en = mid - 1
        }
        mid = (st + en) / 2
    }
    print(mid)
    br.close()
}